cpp/include/raft/interruptible.hpp

```diff
@@ -0,0 +1,246 @@
+   *
+   * @return an object that can be used to cancel the GPU work waited on the given CPU thread.
+   */
+  static auto get_token(std::thread::id thread_id) -> std::shared_ptr<interruptible>
+  {
+    std::lock_guard<std::mutex> guard_get(mutex_);
+    // the following constructs an empty shared_ptr if the key does not exist.
+    auto& weak_store  = registry_[thread_id];
+    auto thread_store = weak_store.lock();
+    if (!thread_store) {
+      thread_store.reset(new interruptible(), [thread_id](auto ts) {
+        std::lock_guard<std::mutex> guard_erase(mutex_);
+        registry_.erase(thread_id);
+        delete ts;
+      });
+      std::weak_ptr<interruptible>(thread_store).swap(weak_store);
+      // Since we're still guarded by the mutex, use this opportunity to initialize the global
+      // cancellation stream. The cancellation stream is only used by non-static cancel(), so it's
+      // safe to initialize it here before the first valid thread store is returned to the user.
+      get_cancellation_stream();
+    }
+    return thread_store;
+  }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by the `thread_id`
+   *
+   * Note, this function uses a mutex to safely get a cancellation token that may be shared
+   * among multiple threads. If you plan to use it from a signal handler, consider the non-static
+   * `cancel_no_throw()` instead.
+   *
+   * @param [in] thread_id a CPU thread, in which the work should be interrupted.
+   *
+   * @throw raft::cuda_error if a CUDA error happens during recording the interruption event.
+   */
+  static inline void cancel(std::thread::id thread_id) { get_token(thread_id)->cancel(); }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by this `interruptible` token.
+   *
+   * Note, this function does not involve thread synchronization/locks, so it should be safe
+   * to call from a signal handler.
+   *
+   * @throw raft::cuda_error if a CUDA error happens during recording the interruption event.
+   */
+  inline void cancel() { RAFT_CUDA_TRY(cancel_no_throw()); }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by this `interruptible` token.
+   *
+   * Note, this function does not involve thread synchronization/locks and not throw any
+   * exceptions, so it's safe to call from a signal handler.
+   *
+   * @return CUDA error code redirected from `cudaEventRecord`.
+   */
+  auto cancel_no_throw() noexcept -> cudaError_t
+  {
+    // This method is supposed to be called from another thread;
+    // multiple calls to it just override each other, and that is ok - the cancellation request
+    // will be delivered (at least once).
+    cancelled_ = true;
+    return cudaEventRecord(wait_interrupt_, get_cancellation_stream());
```

Comment:
I don't understand the purpose of the cancellation stream or recording an event in it when calling `cancel`. So far as I can tell, there is never any synchronization of the cancellation stream nor waiting on the event recorded in the cancellation stream. 

The only time we synchronize with the `wait_interrupt_` event is after it was recorded in the user-provided stream in `synchronize_impl`.

Reply:
The point is when we call `cudaEventRecord(wait_interrupt_, get_cancellation_stream())` from within `cancel` in another CPU thread, the `wait_interrupt_` event can arrive before it would have arrived from `cudaEventRecord(wait_interrupt_, stream)` from within `synchronize` (because user stream is busy, but `get_cancellation_stream()` is usually empty).

Reply:
But when `cudaEventRecord(wait_interrupt_, stream)` is called from within `synchronize`, it will overwrite any state in the `wait_interrupt_` event from a previous `cudaEventRecord(wait_interrupt_, get_cancellation_stream())`. In this way, recording `wait_interrupt_` on `get_cancellation_stream()` can never have side-effects.

Reply:
Hmm, does the overwriting happen at the moment of the call, or at the moment the event is processed in the stream?

In the former case, I think it's working in my experiments and tests, because I always happen to record `wait_interrupt_` on `get_cancellation_stream()` after I do so inside `synchronize`, thus overriding the state again. Then I can just add another check of the `canceled_` flag before the call.

In the latter case, it does not seem to be a problem at all.

Reply:
> Hmm, does the overwriting happen at the moment of the call, or at the moment the event is processed in the stream?

My understanding is that the call to `cudaEventRecord` will overwrite the state.

Reply:
Ok, then adding another `yield_impl()` in [between the `cudaEventRecord` and `cudaEventSynchronize`](https://github.com/rapidsai/raft/blob/b579d72030487f37737ebd9625acc966b3554979/cpp/include/raft/interruptible.hpp#L240-L241) should be enough to consistently detect the interruption.

Reply:
To be honest, I don't think there's any need for an event or cancellation stream whatsoever. Given our discussion above about not being able to call CUDA RT APIs in a signal handler, I think all the signal handler needs to do is to set the `cancelled_` `atomic_flag`. Then the `synchronize_impl_` can be as simple as:
```
  void synchronize_impl(rmm::cuda_stream_view stream)
  {
    yield_impl();
    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
  }
```

At a high level, you just want a synchronize function that checks if the signal handler executed been received before synchronizing the stream. That's exactly what this does.

Reply:
Initially I hoped to use the blocking-wait flag of the event to not use the CPU at 100%. That was the whole reason for using the event instead of a spinning lock. But then I also realized that in some cases (when there are many threads running) it blocked for way too long, so I removed the flag.

Without the blocking-wait, I think, I'd better just implement a spinning lock with `cudaSrteamQuery` like in [comms](https://github.com/rapidsai/raft/blob/b579d72030487f37737ebd9625acc966b3554979/cpp/include/raft/comms/std_comms.hpp#L447-L473) and, as you suggested, use a single `std::atomic_flag` for synchronization. Using `cudaStreamSynchronize` probably is not feasible here, because then the thread cannot be canceled anymore once it arrived at that call.

Thanks for a very thoughtful review!

Reply:
In that case you might want to make `canceled_` be a `std::condition_variable` signaled by the signal handler. Then in `synchronize_impl` you could implement your own version of `cudaStreamSynchronize` with `cudaStreamQuery` and an exponential backoff scheme. Something like:
```
void synchronize_impl(rmm::cuda_stream_view stream)
{
   auto wait = 1us;
   while(not canceled_.wait_for(lock, wait, [stream](){return cudaErrorNotReady != cudaStreamQuery(stream); } ){
      wait *= 2;
   }
}
```

Reply:
But then in the normal case (no cancelling), we'd introduce a latency of up to 100% of the kernel execution time. And we still cannot trigger the condition variable in the signal handler.

Reply:
Ah, of course, can't use condition variables in a signal handler. Just using a lock free atomic and a custom spin w/ backoff is the way to go then.